rholang/src/main/scala/coop/rchain/rholang/interpreter/util/AddressTools.scala

```diff
@@ -16,13 +16,13 @@ class AddressTools(prefix: Array[Byte], keyLength: Int, checksumLength: Int) {
     * @param pk the public key from which the address is derived
     * @return None if the key length is invalid or Some if the address was created successfully
     */
-  def fromPublicKey(pk: PublicKey): Option[String] =
+  def fromPublicKey(pk: PublicKey): Option[Address] =
     if (keyLength == pk.bytes.length) {
       val keyHash = Blake2b256.hash(pk.bytes)
       val payload = prefix ++ keyHash
       val address = payload ++ computeChecksum(payload)
 
-      Some(Base58.encode(address))
+      Some(Address(prefix, address))
```

Comment:
you're putting the whole address (prefix + keyHash + checksum) on the Address.keyHash field here.

Reply:
We should return `Some(Adddress(prefix, keyHash)` and have

```scala
//Address class
def toBase58() = {
    val payload = prefix ++ keyHash
    val address = payload ++ computeChecksum(payload)	      val address = payload ++ computeChecksum(payload)
	address
}
```

Reply:
Done in dev